### ISSUE TITLE
Gridsome Themes

### ISSUE THREAD

**Issue by tomtev** [gridsome/gridsome#484]
This issue exists to discuss how Themes should work in Gridsome.

- How should themes work in a modern framework like Gridsome.
- Are Vue Component libraries with pre-defined layouts better than themes for Gridsome?

Please share your ideas :)

**Comment by atilacamurca** [comment]
Pre-defined layouts, such as [bootstrap-vue](https://bootstrap-vue.js.org) and [vuetify](https://vuetifyjs.com/), are easy to get started and have great documentations. The problem with custom themes is the poor documentation. For instance, I tried to start with the [starter blog theme](https://github.com/gridsome/gridsome-starter-blog/) but as soon as I try to create a new page I was stuck.

Still, I think is valid to create a custom theme, as long as it have some default components and a minimal documentation to help you build something customizable enough.

**Comment by mittalyashu** [comment]
## Questions

* What does this feature request or issue cover?
* What does the theme means? Does it include the design part (color & font) or layouts and website structure too?

I think theming is about using the same naming conventions for different designs, so that while changing from one theme to another theme shouldn't be any problem.

**Comment by jamelt** [comment]
I think theme support is a red herring. I think it's an idea for which there isn't a real use case where you're going to need this.

Having used Hexo for a while now, I've found that _themes_ were a total waste of time and a huge annoyance. Especially, since it needlessly forced to me rearrange my folder structure to accommodate this bogus thing called a theme.

- What is the likelihood that someone would change the entire look and feel of site and expect it to be something you can just swap out without any significant effort?
- This leads me to believe that sites only have one look and feel. And anything else is essentially a rewrite.

**Bottom line:** Don't waste your time implementing this. If someone wants a new _theme_, create a new git branch. There's your theme support.

**Comment by dougouverson** [comment]
I wholeheartedly agree with @jamelt

New tools, new rules! From, "Beyond the Goal: Eliyahu Goldratt Speaks on the Theory of Constraints"

Don't bring over policy baggage from a 15-year-old CMS paradigm!

**Comment by atilacamurca** [comment]
From <https://www.gatsbyjs.org/docs/themes/>

> Using a Gatsby theme, all of your default configuration (shared functionality, data sourcing, design) is abstracted out of your site, and into an installable package.

> This means that the configuration and functionality isn’t directly written into your project, but rather versioned, centrally managed, and installed as a dependency. You can seamlessly update a theme, compose themes together, and even swap out one compatible theme for another.

Maybe we can adapt the ideia to gridsome.

**Comment by Microflash** [comment]
As convenient as themes are for people to get started and swap looks later, I think it is better to have no opinions on this. People are very impatient when they are just getting started (and having a default approach to theming might come in handy at that point) but they become very choosy when they start building their brand which more often than not, doesn't fit into the vision dictated by the framework they are using. The end result: a fight with the framework itself. I have seen this time and again with projects using CSS frameworks and it is always a huge mess and a ton of regret.

In my humble opinion, letting community develop good starters is a better idea than spawning an ecosystem of themes.

**Comment by bjensen** [comment]
Ive been looking around for a good static site generator and since Ive been very happy with Vue I saw this project, however with the lack of themes Ill be using Gatsby since they have starters. Im sure I am not the only one who just wants something that looks pretty up and running within 10 min....

**Comment by tomtev** [comment]
@bjensen Gridsome has starters. Do you mean themes?

**Comment by bjensen** [comment]
Yeah. Sorry. Gatsby calls them starters..

Mvh

Brian
Nordplaner ApS
Fiskerivej 2k, 1.
8000 Aarhus C
+45 60712000

________________________________

**Comment by hacknug** [comment]
@bjensen Gatsby's starters and themes are different things. Like @tomtev said, Gridsome also has starters here: https://gridsome.org/starters/

**Comment by bjensen** [comment]
@hacknug Agreed. I have updated my original post to say themes so there wont be any more confusion. The point is, i think many people like me, will discard this framework quickly because the lack of themes....

**Comment by samuelhorn** [comment]
I would also like something lika themes in packages. Starters are a great way to get started for developers, but once you've started it becomes hard to keep up with updates of the original starter. I would love to make themes people can use and update. I haven't looked that much into Gatsby, but I like the way Jekyll handles themes, I suppose it's kind of the same solution only npm vs gem.

**Comment by jamelt** [comment]
I honestly wouldn't understand why a non-developer would be looking at Gridsome. There are things like WordPress or similar tools that are much geared towards their use case. The fact that Gridsome's big selling point is Vue would make it unappealing to non-developers. 

Also I would add, theme support is something different than having a starter layout that you can build from. Theme support means setting a standard that makes swapping out a look & feel seamlessly over fixed content. This is quite a technical challenges and involves the framework designer making many decisions upfront for all users.

To get a sense of what at theme is, check out CSS Zen Garden (http://www.csszengarden.com/). This site has been around for a very long time, and is an outstanding example of what themes are and how they should function.

**Comment by samuelhorn** [comment]
@jamelt I agree with you on many points. Being a WordPress front-end developer & designer for 10+ years, I totally get that we should not mimic whats called themes in the WordPress community.
And maybe you're right about non developers not going to work with Gridsome. But I still feel that there is a use case for something like what people think about when you talk about themes.

I think of it like if someone, maybe a developer with almost no CSS knowledge is using my starter for a documentation site they're building for their startup. He or she deploys my starter and works with it for a while, maybe changes a color in some kind of override config. Fills it with content and so on. And then, in a year or so, my starter is updated with a lot of new useful stuff. It would be nice if they were able to, in some easy way, adopt all those new changes without the need of manually picking files or update code.

I know this is complex, and definitely should not be a priority right now. But I don't think it hurts to think about it from a lot of different perspectives. 

With that said, again, I agree with much of what you say. And you´re probably right :)

**Comment by Microflash** [comment]
As @jamelt mentioned earlier, creating a support for themes will require having some opinions about how a Gridsome application should be structured. The questions are 
- how flexible these opinions can be, and 
- how easily they can be applied on a given state of a Gridsome application

Angular has got a pretty interesting way to enforce such opinions: **schematics**. I'm not sure what could be a Vue.js counterpart for schematics but something similar can definitely help seed the idea of theming.

**Comment by Hendrixer** [comment]
I feel what Gatsby is doing with themes makes sense. Not so much of a theme in a traditional CMS mind set, but more like composing Gridsome apps together into one. So really good use cases are:

- a headless CMS theme that adds a `/cms-admin` route where your team can edit content
- an e-commerce theme that does something similar to CMS theme but for e-commerce and has routes for products
- an admin dashboard
- a blog
- a hiring / job posts section

All of these would be and should be as simple as `npm i gridsome-theme-ecommerce`.

Me and my company are willing to financially support this.

**Comment by OzzyTheGiant** [comment]
@samuelhorn in my totally biased opinion, I think themes should be a thing of the past. I don't want to give end users any power as to how to control the layout and design of their website because many, if not all of them, don't know how to properly structure a site, let alone give it an excellent design. I think website builders are a joke and that is not what I want Gridsome to ultimately be; I don't want web developers' work to continue being devalued because of cheap themes being sold out there. Instead we should focus on building component libraries similar to Vuetify or even on top of existing ones. This way, the control remains in the hands of the developer while drastically reducing development time, allowing developers to charge however much they want (within reason).

**Comment by gkweb** [comment]
I am against themes being baked into the framework. I do like the idea of shareable and maintainable abstractions though. Provided the setup is simple it would be a neat feature. This wouldn't be my deciding factor on using the Grisome though. The first thing I'm looking at with Vue projects right now is good composition api and Vue 3 support.
